### PR TITLE
[dynamo exporter] Support string in dynamic_shapes

### PR DESCRIPTION
**Pull request by titaiwangms**
## Describe your changes

With https://github.com/pytorch/pytorch/pull/146321, `torch.onnx.export(.., dynamo=True)` now can support string in dynamic_shapes, which fits better with Olive driven with configuration.

Major changes:
- Add support for string in dynamic_shapes
    - Move dynamic_shapes pre-process to io_config.py (like dynamic_axes)
    - Get rid of the post-process of making [str, int, int] -> torch.export.Dim(str, max=int, min=int). `torch.onn.export(..., dynamo=True)` can now take string.
    - Leverage Optimum to auto-generate dynamic_shapes when Optimum models is requested.
    - KV cache support

Pitfall:
- When dynamic_shapes targets kwargs, both of them need to follow the order of model.forward signature. onnx/conversion.py provides naive approach to sort them, but users should be aware of this.

Minor changes:
- Move onnxscript (released) to the official requirement.txt
- dynamic_shapes with string is supported since torch 2.7

## Checklist before requesting a review
- [x] Add unit tests for this change.
- [ ] Make sure all tests can pass.
- [ ] Update documents if necessary.
- [x] Lint and apply fixes to your code by running `lintrunner -a`
- [x] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.
- [ ] Is this PR including examples changes? If yes, please remember to update [example documentation](https://github.com/microsoft/Olive/blob/main/docs/source/examples.md) in a follow-up PR.

### CODE REVIEW

**Comment by titaiwangms** [comment]
Some CI tests are failing with docker issues.

File: olive/common/hf/model_io.py [hunk 1]
```diff
@@ -124,10 +125,59 @@ def get_model_io_config(model_name: str, task: str, model: "PreTrainedModel", **
...
+            # From Optimum: past_key_values.0.key, past_key_values.0.value,
+            #               past_key_values.1.key, past_key_values.1.value, ...
+            idx = int(input_name.split(".")[1])
+            max_idx = max(max_idx, idx)
+            past_key_value_count += 1
+
+    # Check if we have exactly 2 * (max_idx + 1) key-value pairs
+    expected_count = 2 * (max_idx + 1)
+    if past_key_value_count != expected_count or past_key_value_count % 2 != 0:
+        warnings.warn(
```
Comment (jambayk) [inline olive/common/hf/model_io.py:155]: nit: can you use logger.warn?

File: olive/passes/onnx/conversion.py [hunk 2]
```diff
@@ -214,15 +217,22 @@ def _export_pytorch_model(
         if config.use_dynamo_exporter:
             # Take the "release" version so that dev builds like 2.5.0dev1234 are treated as 2.5.0
             torch_version = version.parse(torch.__version__).release
+            if torch_version < version.parse("2.7.0").release:
+                warnings.warn(
```
Comment (jambayk) [inline olive/passes/onnx/conversion.py:221]: logger.warning. 

Should we add a condition for `io_config.dynamic_shapes is not None`? If "dynamic" is false or user didn't provide dynamic shapes, then this warning is not needed

File: olive/passes/onnx/conversion.py [hunk 3]
```diff
@@ -601,88 +612,30 @@ def _validate_dynamic_shapes(dynamic_shapes, dummy_inputs):
...
+        # Sort io_config.dynamic_shapes based on this order
+        unflatten_dynamic_shapes = collections.OrderedDict(
+            sorted(unflatten_dynamic_shapes.items(), key=lambda item: param_order.index(item[0]))
+        )
+        dummy_inputs = collections.OrderedDict(
+            sorted(dummy_inputs.items(), key=lambda item: param_order.index(item[0]))
+        )
+        # dummy_inputs is kwargs
+        return unflatten_dynamic_shapes, (), dummy_inputs
+    # If dynami_shapes and dummy_inputs are both list/tuple, we don't need to do anything.
```
Comment (jambayk) [inline olive/passes/onnx/conversion.py:639]: nit: typo

**Comment by titaiwangms** [comment]
Replaced by #1639
